File: backend/src/main/java/com/sigtm/backend/cliente/ClienteService.java
package com.sigtm.backend.cliente;

import com.sigtm.backend.cliente.dto.ClienteUpdateRequest;
import com.sigtm.backend.common.ResourceNotFoundException;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;

@Service
public class ClienteService {

    private final ClienteRepository clienteRepository;

    public ClienteService(ClienteRepository clienteRepository) {
        this.clienteRepository = clienteRepository;
    }

    /** Solo el rol TALLER puede listar todos los clientes (se valida en el controller via SecurityConfig). */
    public List<Cliente> listarTodos() {
        return clienteRepository.findAll();
    }

    public Cliente obtenerPorId(Long id) {
        return clienteRepository.findById(id)
                .orElseThrow(() -> new ResourceNotFoundException("Cliente no encontrado: " + id));
    }

    public Cliente obtenerPorUsuarioId(Long usuarioId) {
        return clienteRepository.findByUsuarioId(usuarioId)
                .orElseThrow(() -> new ResourceNotFoundException("No existe un cliente asociado a este usuario."));
    }

    @Transactional
    public Cliente actualizar(Long id, ClienteUpdateRequest request) {
        Cliente cliente = obtenerPorId(id);
        cliente.setNombre(request.nombre());
        cliente.setTelefono(request.telefono());
        cliente.setDocumento(request.documento());
        return clienteRepository.save(cliente);
    }
}

File: src/services/cliente.service.js
const clienteRepository = require('../repositories/cliente.repository');
const usuarioRepository = require('../repositories/usuario.repository');
const AppError = require('../utils/AppError');

async function crear({ usuarioId, telefono, direccion }) {
  const usuario = await usuarioRepository.buscarPorId(usuarioId);
  if (!usuario) {
    throw new AppError('El usuario asociado no existe.', 404);
  }

  const clienteExistente = await clienteRepository.buscarPorUsuarioId(usuarioId);
  if (clienteExistente) {
    throw new AppError('Este usuario ya tiene un perfil de cliente.', 409);
  }

  return clienteRepository.crear({ usuarioId, telefono, direccion });
}

async function listar() {
  return clienteRepository.listar();
}

async function obtenerPorId(id) {
  const cliente = await clienteRepository.buscarPorId(id);
  if (!cliente) {
    throw new AppError('Cliente no encontrado.', 404);
  }
  return cliente;
}


async function obtenerPorUsuarioId(usuarioId) {
  const cliente = await clienteRepository.buscarPorUsuarioId(usuarioId);
  if (!cliente) {
    throw new AppError('No se encontro un perfil de cliente para este usuario.', 404);
  }
  return cliente;
}

async function actualizar(id, datos) {
  await obtenerPorId(id); 
  return clienteRepository.actualizar(id, datos);
}

module.exports = { crear, listar, obtenerPorId, obtenerPorUsuarioId, actualizar };